### DISCUSSION TITLE
Help getting reference highlighting working

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1326] [Q&A] [answered]
[Related](https://github.com/nvim-treesitter/nvim-treesitter/discussions/1667), although in this case I can't get it to work with the `tree-sitter highlight` command. I'm using tree-sitter-cli 0.20.0 for generation of the grammar, and tree-sitter-cli 0.19.5 for running the highlight command.

I've been writing highlights for [my tree-sitter grammar](https://github.com/tlaplus-community/tree-sitter-tlaplus/tree/ts-highlight), just basic stuff like `(rulename) @capturename`. All is well. Now I'm trying to get definitions & references working and am running into trouble. Consider a simple TLA+ module like:
```tla
---- MODULE Test ----
CONSTANT Foo
op == Foo
f(x) == x
====
```
which declares a global constant, `Foo`, then a definition containing a reference to that constant, then a definition taking a parameter. I would like the identifier references to be highlighted the same as their definitions. The tree-sitter parse tree for this module is:
```
(source_file
  (module (single_line) name: (identifier) (single_line)
    (unit (constant_declaration (identifier)))     
    (unit (operator_definition
      name: (identifier)
      (def_eq)
      definition: (identifier_ref)
    ))
    (unit (operator_definition
      name: (identifier)
      parameter: (identifier)
      (def_eq)
      definition: (identifier_ref)
    ))
  (double_line))
)
```
I can highlight `Foo` and the `x` in `f(x)` with the following queries in my highlights.scm file:
```
(constant_declaration (identifier) @constant)
(operator_definition parameter: (identifier) @variable.parameter)
```
Here's the `tree-sitter highlight` output; see that `Foo` is colored brown and `x` is underlined:
![image](https://user-images.githubusercontent.com/718765/128568150-5d24d096-bc85-4d47-9284-a25f07037591.png)

I want to highlight the `identifier_ref` nodes appropriately. I've tried the following in my locals.scm file:
```
(module) @local.scope
(operator_definition) @local.scope

(constant_declaration (identifier) @local.definition)
(operator_definition parameter: (identifier) @local.definition)

(identifier_ref) @local.reference
```
and have this in my `package.json`:
```
"tree-sitter": [
  {
    "scope": "source.tlaplus",
    "file-types": [
      "tla"
    ],
    "highlights": [
      "queries/highlights.scm"
    ],
    "locals": [
      "queries/locals.scm"
    ]
  }
]
```
but the `tree-sitter highlight` output is unchanged:
![image](https://user-images.githubusercontent.com/718765/128568150-5d24d096-bc85-4d47-9284-a25f07037591.png)
Am I doing something obviously wrong?

**Comment by ahlinc** [comment, marked as answer]
@ahelwer You need to specify in the `highlights.scm` file that your `identifier_ref` nodes should be also highlighted. Definitions in the `locals.scm` just serve for color assignments coordination but don't defining that something should be colored.

<details><summary><i>Demo:</i></summary>

![Screenshot from 2021-08-19 08-30-33](https://user-images.githubusercontent.com/14666676/130013212-30eebad6-0cc0-45bc-bdbd-2e30b94bf989.png)

```diff
diff --git i/queries/highlights.scm w/queries/highlights.scm
index 7d3f885..4c6e4ea 100644
--- i/queries/highlights.scm
+++ w/queries/highlights.scm
@@ -103,6 +103,9 @@
 (function_definition (quantifier_bound (tuple_of_identifiers (identifier) @variable.parameter)))
 (lambda (identifier) @variable.parameter)

+; References
+(identifier_ref) @variable.reference
+
 ; Delimiters
 [
   (langle_bracket)
```

</details>

**Reply by ahelwer** [reply]
Excellent thank you, that works great! I believe there is an oversight in https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables documentation where it doesn't mention that the identifier references must have a default type of highlighting for the reference-based highlighting to trigger.
